src/chains/ethereum/console.log/scripts/helpers.ts

```diff
@@ -0,0 +1,337 @@
+    // match. We only want to throw if the signatures themselves differ while
+    // the 4-bytes match.
+    if (signatureCache.get(signatureInt) !== abiSignatureString) {
+      // if we've already generated this signature before throw!
+      throw new Error(
+        `Signature collision detected between log(${abiSignatureString}) and ???(${signatureCache.get(
+          signatureInt
+        )})`
+      );
+    }
+  }
+  signatureCache.set(signatureInt, abiSignatureString);
+
+  const names =
+    params.length === 1 ? ["value"] : params.map((_, i) => `value${i + 1}`);
+  const fullParamsWithNames = params.map((arg, i) => arg + " " + names[i]);
+  const encodeArgs = [`"${abiSignatureString}"`, ...names];
+  const printComment = params.length > 1 && abiParams[0] === "string";
+
+  const solidity = `${printComment ? COMMENT : ""}
+    function ${solidityFunctionName}(${fullParamsWithNames.join(
+    ", "
+  )}) internal view {
+        _sendLogPayload(abi.encodeWithSignature(${encodeArgs.join(", ")}));
+    }`;
+
+  const javascriptHandlers = params.map(arg => typeToHandlerMap.get(arg));
+  const javascript = `  // ${abiSignatureString}
+  [${signatureInt}, [${javascriptHandlers.join(", ")}]]`;
+
+  return {
+    solidity,
+    javascript,
+    params,
+    name: solidityFunctionName
+  };
+}
+
+function getNamedLogFunctionName(type: SolidityType) {
+  const logName = `log${
+    type[0].toUpperCase() + type.replace(" memory", "").slice(1)
+  }`;
+  return logName;
+}
+
+/**
+ * Generates signature code as `logString(string memory)`or `logAddress(address)`
+ * instead of just `log(string memory)` or `log(address)`.
+ *
+ * Might yield multiple signatures, like `"logUint256"` and `"logUint"` for
+ * `solidityType` `"uint256"`.
+ *
+ * @param solidityType
+ */
+function* getNamedSignature(solidityType: SolidityType) {
+  const logName = getNamedLogFunctionName(solidityType);
+  yield getSignature([solidityType], logName);
+
+  if (hardhatTypeAliases.get(solidityType)) {
+    const alias = hardhatTypeAliases.get(solidityType);
+    const aliasLogName = getNamedLogFunctionName(alias);
+    const solditiySignature = getSignature([solidityType], aliasLogName);
+    delete solditiySignature.javascript;
+    const javascriptSignature = getSignature([alias], aliasLogName);
+    delete javascriptSignature.solidity;
+    yield solditiySignature;
+    yield javascriptSignature;
+  }
+}
+
+/**
+ * Combines the array of types into every permutation of types from length 1 to
+ * array.length.
+ * @param array
+ */
+function* permute<T>(array: T[] | readonly T[]) {
+  for (let i = 0; i < array.length; i++) {
+    const length = Math.pow(array.length, i + 1);
+    for (let j = 0; j < length; j++) {
+      const parameters: SolidityType[] = [];
+
+      for (let k = i; k >= 0; k--) {
+        const denominator = Math.pow(array.length, k);
+        const index = Math.floor(j / denominator) % combinatorTypes.length;
```

Comment:
```suggestion
        const index = Math.floor(j / denominator) % array.length;
```
The only use of `permute` passes in `combinatorTypes`, but we use `array.length` everywhere else. Is there a reason to use `combinatorTypes.length` here?

Reply:
Good catch. This was just a bug waiting to happen!